dandi/download.py

```diff
@@ -1,121 +1,172 @@
 from .utils import flatten, flattened, Parallel, delayed
 
 lgr = get_logger()
 
 
-def parse_dandi_url(url):
-    """Parse url like and return server (address), asset_id and/or directory
-
-    Example URLs (as of 20200310):
-    - User public: (Users -> bendichter/Public/Tolias2020)
-      [seems to be visible only if logged in]
-      https://gui.dandiarchive.org/#/folder/5e5593cc1a343161ff7c5a92
-      https://girder.dandiarchive.org/#user/5da4b8fe51c340795cb18fd0/folder/5e5593cc1a343161ff7c5a92
-    - Collection top level (Collections -> yarik):
-      https://gui.dandiarchive.org/#/collection/5daa5ca7e3489855a3027682
-      https://girder.dandiarchive.org/#collection/5daa5ca7e3489855a3027682
-    - Collections: (Collections -> yarik/svoboda)
-      https://gui.dandiarchive.org/#/folder/5dab0830f377535c7d96c2b4
-      https://girder.dandiarchive.org/#collection/5daa5ca7e3489855a3027682/folder/5dab0830f377535c7d96c2b4
-    - Dataset landing page metadata
-      https://gui.dandiarchive.org/#/dandiset-meta/5e6d5c6976569eb93f451e4f
-
-    Individual and multiple files:
-      - dandi???
-      - girder -- we don't support:
-        https://girder.dandiarchive.org/api/v1/item/5dab0972f377535c7d96c392/download
-      - gui.: support single or multiple
-        # if there is a selection, we could get multiple items
-        https://gui.dandiarchive.org/#/folder/5e60c14f81bc3e47d94aa012/selected/item+5e60c19381bc3e47d94aa014
-
-    Multiple selected files + folders -- we do not support ATM, then further
-    RFing would be due, probably making this into a generator or returning a
-    list of entries.
-
-    "Features":
-
-    - supports DANDI naming, such as https://dandiarchive.org/dandiset/000001
-      Since currently redirects, it just resorts to redirect if url lacks #.
-      TODO: make more efficient, .head instead of .get or some other way to avoid
-      full download_file.
-    - uses some of `known_instance`s to map some urls, e.g. from
-      gui.dandiarchive.org ones into girder.
-
-    Returns
-    -------
-    server, asset_type, asset_id
-      asset_type is either asset_id or folder ATM. asset_id might be a list
-      in case of multiple files
-
-    """
-    if "#" not in url:
+class _dandi_url_parser:
+    # Defining as a class with all the attributes to not leak all the variables etc
+    # into module space, and later we might end up with classes for those anyways
+    id_regex = "[a-f0-9]{24}"
+    id_grp = f"(?P<id>{id_regex})"
+    server_grp = "(?P<server>(?P<protocol>https?)://(?P<hostname>[^/]+)/)"
+    known_urls = {
+        # Those we first redirect and then handle the redirected URL
+        # TODO: Later should better conform to our API, so we could allow
+        #       for not only "dandiarchive.org" URLs
+        "https?://dandiarchive.org/.*": {"handle_redirect": True},
+        "https?://[^/]*dandiarchive-org.netlify.app/.*": {"map_instance": "dandi"},
```

Comment:
it seems odd to support netlify, as these domains are not part of our production infrastructure. i'm guessing this makes it easier for you to test/interact with the previews?

Reply:
for testing dev/master previews, should we deploy a version of it on github simultaneously (through github actions)? then at least it could be reduced to `https://*dandiarchive.org/.*`

Reply:
In principle, I could may be make it into development option - available only if DANDI_DEVEL set. But since this is not visible to the user anyways, be it wouldn't really help anything IMHO.
The rationale for adding it - being able to test client on a proposed Ina PR deployment. If we somehow could, it would've been nice to even automate it in dandiarchive tests - feed client a few urls archive is "producing" in its pr state.

Reply:
As for an alternative deployment - if way to do it easily per or then sure why not, but then why do we use netlify?

Reply:
netlfiy allows us to preview any PR easily. while that can be done with github actions, it's a bit more complicated because of the workflow. github actions is replacing what a lot of people used netlify/travis/circle for.

Reply:
ok, then for now let's proceed as is ;-)

Reply:
In the future, I'd like to be able to get away from URLs as an API between the front end and the CLI. I'd prefer something more like where the web UI tells you "use this command to download this dandiset: `dandi download 000023`".

Reply:
In the future, I'd prefer web ui to use our API urls ;-)